opentreemap/treemap/instance.py

```diff
@@ -768,9 +771,39 @@ def raise_errors(errors):
         if errors:
             raise_errors(errors)
 
-    def save(self, *args, **kwargs):
+    @transaction.atomic
+    def _bulk_assign_task_reference_numbers(self, tasks):
+        instance = Instance.objects.select_for_update().get(id=self.id)
+        for task in tasks:
+            task.reference_number = instance.task_sequence_number
+            instance.task_sequence_number += 1
+        instance.save(update_fields=['task_sequence_number'])
+
+    @transaction.atomic
+    def _bulk_assign_work_order_reference_numbers(self, work_orders):
+        instance = Instance.objects.select_for_update().get(id=self.id)
+        for work_order in work_orders:
+            work_order.reference_number = instance.work_order_sequence_number
+            instance.work_order_sequence_number += 1
+        instance.save(update_fields=['work_order_sequence_number'])
+
+    def save(self, tasks=None, work_orders=None, *args, **kwargs):
+        """
+        Keyword Arguments:
+
+        tasks - List of Task models
+        work_orders - List of WorkOrder models
+        """
         self.full_clean()
 
         self.url_name = self.url_name.lower()
 
+        if tasks:
```

Comment:
Based on the small sample of potential customers we have talked to multiple people creating work orders at the same time would not be a common occurrence. In the interest of standing up the feature sooner rather than later, we could defer integrity error protection.

I was unable to find documentation on Django unmanaged fields. I could only find documentation on unmanaged models.

I think the "unmanaged field" philosophy is correct, though. The sequence numbers will always need to be fetched and updated directly through queryset functions, not via the `Instance` model properties.

I don't think we need to involve another datastore (Redis) to solve the problem, just change the way we are interacting with the Postgres database in this situation.